core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala

```diff
@@ -0,0 +1,191 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package kafka.server.epoch
+
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kafka.server.LogOffsetMetadata
+import kafka.server.checkpoints.LeaderEpochCheckpoint
+import kafka.server.epoch.Constants.{UNSUPPORTED_EPOCH, UNSUPPORTED_EPOCH_OFFSET}
+import kafka.utils.CoreUtils._
+import kafka.utils.{Logging}
+import org.apache.kafka.common.requests.EpochEndOffset
+import scala.collection.mutable.ListBuffer
+
+trait LeaderEpochCache {
+  def assignToLeo(leaderEpoch: Int)
+  def assign(leaderEpoch: Int, offset: Long)
+  def latestEpoch(): Int
+  def endOffsetFor(epoch: Int): Long
+  def clearLatest(offset: Long, retainMatchingOffset: Boolean = true)
+  def clearOldest(offset: Long, retainMatchingOffset: Boolean = true)
+  def clear()
+}
+
+object Constants {
+  val UNSUPPORTED_EPOCH_OFFSET = EpochEndOffset.UNDEFINED_OFFSET
+  val UNSUPPORTED_EPOCH = -1
+}
+
+/**
+  * Represents a cache of (LeaderEpoch => Offset) mappings for a particular replica.
+  *
+  * Leader Epoch = epoch assigned to each leader by the controller.
+  * Offset = offset of the first message in each epoch.
+  *
+  * @param leo a function that determines the log end offset
+  * @param checkpoint the checkpoint file
+  */
+class LeaderEpochFileCache(leo: () => LogOffsetMetadata, checkpoint: LeaderEpochCheckpoint) extends LeaderEpochCache with Logging {
+  private val lock = new ReentrantReadWriteLock()
+  private[epoch] var epochs = lock synchronized { ListBuffer(checkpoint.read(): _*) }
+
+  /**
+    * Assigns the supplied Leader Epoch to the current LEO
+    * Once the epoch is assigned it cannot be reassigned
+    *
+    * @param epoch
+    */
+  override def assignToLeo(epoch: Int) = {
+    assign(epoch, leo().messageOffset)
+  }
+
+  /**
+    * Assigns the supplied Leader Epoch to the supplied Offset
+    * Once the epoch is assigned it cannot be reassigned
+    *
+    * @param epoch
+    * @param offset
+    */
+  override def assign(epoch: Int, offset: Long): Unit = {
+    inWriteLock(lock) {
+      if (epoch >= 0 && epoch > latestEpoch()) {
```

Comment:
Have changed to: `offset >= the offset of the last epoch.` as epoch can increment on the leader when the offset does not change.